src/Configuration/Programmer/Programmer_cmd.cpp

```diff
@@ -19,57 +19,155 @@ You should have received a copy of the GNU General Public License
-      CFG_POST_MSG("Device name: %s", device.name.c_str());
-      CFG_POST_MSG("Device index: %s", std::to_string(device.index).c_str());
-      CFG_POST_MSG("Device jtagId: %s", device.jtagId.c_str());
-      CFG_POST_MSG("Device mask: %s", std::to_string(device.mask).c_str());
-      CFG_POST_MSG("Device irlength: %s",
-                   std::to_string(device.irlength).c_str());
-      CFG_POST_MSG("Device flashSize: %s",
-                   std::to_string(device.flashSize).c_str());
+  // It is currently to use for
+  // programmer API manual testing
+  // <TODO>
+  // eventually it will be a standalone programmer executable
+
+  auto formatMemorySize = [](int sizeInBits) -> std::string {
+    const int bitsInKilobit = 1024;
+    const int bitsInMegabit = 1024 * 1024;
+
+    if (sizeInBits >= bitsInMegabit) {
+      int megabits = sizeInBits / bitsInMegabit;
+      return std::to_string(megabits) + "M";
+    } else if (sizeInBits >= bitsInKilobit) {
+      int kilobits = sizeInBits / bitsInKilobit;
+      return std::to_string(kilobits) + "K";
+    } else {
+      return std::to_string(sizeInBits);
     }
+  };
+
+  // openocd path is the same as the executable path
+  std::filesystem::path openOcdExecutablePath = std::filesystem::current_path();
+  std::string openOcdName;
+// if windows
+#ifdef _WIN32
+  openOcdName = "openocd.exe";
+#else
+  openOcdName = "openocd";
+#endif
+  openOcdExecutablePath /= openOcdName;
+
+  int ret = InitLibrary(openOcdExecutablePath.string());
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+    return ret;
+  }
+  std::vector<Device> devices;
+
+  std::vector<FOEDAG::Cable> cables;
+  ret = GetAvailableCables(cables);
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+    return ret;
+  } else if (cables.size() == 0) {
+    CFG_POST_MSG("No programming cable found. Make sure it is connected.");
+    return -1;
+  }
+  CFG_POST_MSG("return %d", ret);
+  CFG_POST_MSG("-- GetAvailableCables(cables) --")
+  for (auto& cable : cables) {
+    CFG_POST_MSG("--------------")
+    CFG_POST_MSG("product ID: 0x%04x", cable.productId);
+    CFG_POST_MSG("vendor  ID: 0x%04x", cable.vendorId);
+    CFG_POST_MSG("serial    : %s", cable.serialNumber.c_str());
+    CFG_POST_MSG("desc      : %s", cable.description.c_str());
+    CFG_POST_MSG("bus       : %d", cable.busAddr);
+    CFG_POST_MSG("port      : %d", cable.portAddr);
+    CFG_POST_MSG("device    : %d", cable.deviceAddr);
+    CFG_POST_MSG("speed     : %d", cable.speed);
+    CFG_POST_MSG("transport : %d", cable.transport);
+    CFG_POST_MSG("channel   : %d", cable.channel);
+  }
+
+  CFG_POST_MSG("-- ListDevices(cables, outputMsg) --")
+  ret = ListDevices(cables[0], devices);
+  if (ret != NoError) {
+    CFG_POST_MSG("Error code = %d. %s", ret, GetErrorMessage(ret).c_str());
+    return ret;
+  }
+  if (devices.size() <= 0) {
```

Comment:
strictly speaking the checking should be "=="

size() for vector/list is size_t which is an unsigned number. neg is invalid.

Reply:
Ok. I will fix it to use '=='